rmf_task_ros2/src/rmf_task_ros2/dispatcher/Dispatcher.cpp

```diff
@@ -0,0 +1,323 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include <rmf_task_ros2/dispatcher/Dispatcher.hpp>
+
+#include <rclcpp/node.hpp>
+
+namespace rmf_task_ros2 {
+namespace dispatcher {
+
+using DispatchRequest = rmf_task_msgs::msg::DispatchRequest;
+using TaskSummary = rmf_task_msgs::msg::TaskSummary;
+
+//==============================================================================
+class Dispatcher::Implementation
+{
+public:
+  std::shared_ptr<rclcpp::Node> node;
+  std::shared_ptr<bidding::Auctioneer> auctioneer;
+  std::shared_ptr<action::TaskActionClient<DispatchRequest,
+    TaskSummary>> action_client;
+
+  StatusCallback on_change_fn;
+
+  DispatchTasksPtr active_dispatch_tasks;
+  DispatchTasksPtr terminal_dispatch_tasks;
+  int i = 0; // temp index for generating task_id
+  double bidding_time_window;
+
+  Implementation(std::shared_ptr<rclcpp::Node> node_)
+  : node(std::move(node_))
+  {
+    active_dispatch_tasks = std::make_shared<DispatchTasks>();
+    terminal_dispatch_tasks = std::make_shared<DispatchTasks>();
+
+    // ros2 param
+    bidding_time_window =
+      node->declare_parameter<double>("bidding_time_window", 2.0);
+    RCLCPP_WARN(node->get_logger(),
+      " Declared Time Window Param as: %f secs", bidding_time_window);
+  }
+
+  void start()
+  {
+    using namespace std::placeholders;
+    auctioneer->receive_bidding_result(
+      std::bind(&Implementation::receive_bidding_winner_cb, this, _1, _2));
+    action_client->on_terminate(
+      std::bind(&Implementation::terminate_task, this, _1));
+    action_client->on_change(
+      std::bind(&Implementation::task_status_cb, this, _1));
+  }
+
+  TaskID submit_task(const TaskProfile& task)
+  {
+    auto submitted_task = task;
+
+    // auto generate a task_id for a given submitted task
+    // TODO: generate a unique task_id based of clock
+    submitted_task.task_id =
+      std::to_string( ((int)task.task_type.type)*1000 + (i++) );
```

Comment:
Currently, the dispatcher is working as the "centralized task server", which will generate an ID in response during submit_task().  I admit that it seems confusing [here](https://github.com/osrf/rmf_core/blob/ec4e9436e1e3f56a238cd2cdcba924004ab4e60f/rmf_task_ros2/include/rmf_task_ros2/dispatcher/Dispatcher.hpp#L72) as I wouldn't want to introduce another variable like this. Open for amendment.
On the other hand, we are generating the `task_id` in this manner for easy debugging, do you have any advice on this?

Reply:
I guess what confuses me most is that `TaskProfile` has a `task_id` field. This function is receiving a `TaskProfile` and then completely ignoring its `task_id` and generating a brand new one. To me, this indicates some inconsistency in responsibilities.

Doing the message breakdown and receiving a `TaskDescription` instead of a `TaskProfile` would be preferable in my mind.

> On the other hand, we are generating the task_id in this manner for easy debugging, do you have any advice on this?

I think for the sake of human-readability, the best `task_id` would be `task_type_name + std::to_string(task_counter++)` where `task_type_name` is the name of the task type, in characters. This will require making a `std::vector<std::string>` that maps from task type index to a string, which is a bit of a chore to maintain, but I think it's well worth the readability.

On a separate note, we should never use the C-style cast `(int)task.task_type.type`. Instead we should always use the safe C++ cast: `static_cast<std::size_t>(task.task_type.type)`. It's more verbose, but it can help avoid disastrous Undefined Behavior.

Reply:
Good point. This introduces a new msg (`TaskDescription.msg`). Changes: https://github.com/osrf/rmf_core/pull/228/commits/a82ec9d6bc574b80edab3635bd62c8fe89c8288d https://github.com/osrf/rmf_core/pull/228/commits/8a00de7147d312c006ce23e24a424bc654ab7af1